ipa-core/src/protocol/context/mod.rs

```diff
@@ -333,38 +362,170 @@ impl<'a> Inner<'a> {
+    let rcv_stream = ctx
+        .recv_from_shards::<K>()
+        .map(|(shard_id, v)| {
+            (
+                shard_id,
+                v.map(Option::Some).map_err(crate::error::Error::from),
+            )
+        })
+        .fuse();
+
+    // This produces a stream of outcomes for each se
+    // In order to make it compatible with receive stream, it also returns records that must
+    // stay on this shard, according to `shard_picker`'s decision.
+    // That gives an awkward interface for output: (destination shard, Result<Option<Value>>)
+    // The second argument is set to Err if we failed to send the value out. This will fail the
+    // whole resharding process.
+    // If send was successful, we set the argument to Ok(None). Only records assigned to this shard
+    // by the `shard_picker` will have the value of Ok(Some(Value))
+    let send_stream = futures::stream::unfold(
+        // it is crucial that the following execution is completed sequentially, in order for record id
+        // tracking per shard to work correctly. If tasks complete out of order, this will cause share
+        // misplacement on the recipient side.
+        (
+            input.enumerate().zip(iter::repeat(ctx.clone())),
+            &mut sending_ends,
+        ),
+        |(mut input, shard_ends)| async {
+            // Process more data as it comes in, or close the sending channels, if there is nothing
+            // left.
+            if let Some(((i, val), ctx)) = input.next() {
+                let dest_shard = shard_picker(ctx, RecordId::from(i), &val);
+                if dest_shard == my_shard {
+                    Some(((my_shard, Ok(Some(val.clone()))), (input, shard_ends)))
+                } else {
+                    let (record_id, se) = shard_ends.get_mut(&dest_shard).unwrap();
+                    let send_result = se
+                        .send(*record_id, val)
+                        .await
+                        .map_err(crate::error::Error::from)
+                        .map(|()| None);
+                    *record_id += 1;
+                    Some(((my_shard, send_result), (input, shard_ends)))
+                }
+            } else {
+                for (closing_id, sending_end) in shard_ends.values() {
+                    sending_end.close(*closing_id).await;
+                }
+                None
+            }
+        },
+    )
+    .fuse();
+
+    // This contains the deterministic order of events after resharding is complete.
+    // Each shard will hold the records in this order:
+    // [shard_0_records], [shard_1_records], ..., [shard_N].
+    // There is no reason why this strategy was chosen. As long as it is consistent across helpers,
+    // other ways to build the total order work too. For example, we could put records with
+    // record_id = 0 first, then records with record_id = 1, etc.
+    let mut r: Vec<Vec<_>> = ctx.shard_count().iter().map(|_| Vec::new()).collect();
```

Comment:
Maybe use `Vec::with_capacity` here? Even a conservative (for large inputs) bound like `1.1 * input_len / shard_count` seems like a win.